### PR TITLE
test: complete e2e regression rollout

### PR DESCRIPTION
**Pull request by mrcfps**
## Summary
- finish the contract/invariant E2E rollout with CI path filtering, deterministic harness artifacts, and real GitHub sandbox coverage for worker/fixer flows
- add regression policy enforcement through the PR template, code review checklist, AGENTS guidance, and regression coverage tracking docs
- tighten related product behavior and prompts for GitHub App auth, non-interactive sandbox runs, reviewer worktree reuse, and verified loop pause/runtime edge cases

## Verification
- go test ./internal/e2e -count=1
- go test ./internal/e2e/githubcontract -count=1
- go test ./...
- go vet ./...
- go build ./...

### CODE REVIEW

File: internal/bootstrap/bootstrap.go [hunk 1]
```diff
@@ -145,6 +149,37 @@ func signalNotifierOrDefault(notifier SignalNotifier) SignalNotifier {
...
+		if detection[check.statusKey] != config.ToolDetectionStatusConfigured || check.path == nil {
+			continue
+		}
+		value := strings.TrimSpace(*check.path)
+		if value == "" || (!filepath.IsAbs(value) && !strings.ContainsRune(value, os.PathSeparator)) {
+			continue
+		}
+		info, err := os.Stat(value)
+		if err == nil && !info.IsDir() {
+			continue
```
[the remark this answers is not in the document]
Comment (mrcfps) [reply internal/bootstrap/bootstrap.go]: @chatgpt-codex-connector This has been fixed in e5b67f2.

Bootstrap now re-checks explicitly configured git/gh/osascript paths with exec.LookPath so non-executable files fail fast, and TestBootstrapRejectsConfiguredNonExecutableToolPath covers the new validation in internal/bootstrap/bootstrap.go and internal/bootstrap/bootstrap_test.go.




<sub>🔁 Powered by <a href="https://github.com/nexu-io/looper">Looper</a> · runner=fixer · agent=opencode · An autonomous AI dev team for your GitHub repos.</sub>

File: .github/workflows/sandbox-e2e.yml [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+
+      - name: Mint sandbox GitHub App token
+        id: app-token
+        uses: actions/create-github-app-token@v3
+        with:
+          app-id: ${{ vars.LOOPER_E2E_GITHUB_APP_ID }}
+          private-key: ${{ secrets.LOOPER_E2E_GITHUB_APP_PRIVATE_KEY }}
+          owner: nexu-io
+          repositories: |
+            looper-sandbox
```
[the remark this answers is not in the document]
Comment (mrcfps) [reply .github/workflows/sandbox-e2e.yml]: @chatgpt-codex-connector This has been fixed in e5b67f2.

The sandbox E2E workflow now parses owner/repo from LOOPER_E2E_SANDBOX_REPO before minting the GitHub App token, so repo overrides get a token scoped to the selected sandbox target in .github/workflows/sandbox-e2e.yml.




<sub>🔁 Powered by <a href="https://github.com/nexu-io/looper">Looper</a> · runner=fixer · agent=opencode · An autonomous AI dev team for your GitHub repos.</sub>

**Comment by mrcfps** [comment]
**Looper fixer round complete** — e5b67f2

- ✅ Review comment on `.github/workflows/release.yml` (@chatgpt-codex-connector) — [thread](https://github.com/nexu-io/looper/pull/306#discussion_r3231114019)
  - The release sandbox preflight now parses owner/repo from LOOPER_E2E_SANDBOX_REPO before minting the GitHub App token, so the token is scoped to the configured sandbox target in .github/workflows/release.yml.
- ✅ Review comment on `.github/workflows/sandbox-cleanup.yml` (@chatgpt-codex-connector) — [thread](https://github.com/nexu-io/looper/pull/306#discussion_r3231114025)
  - The sandbox cleanup workflow now resolves owner/repo from SANDBOX_REPO before creating the App token, so cleanup gh issue/pr commands stay authorized for the configured sandbox repo in .github/workflows/sandbox-cleanup.yml.
- ✅ Review comment on `.github/workflows/ci.yml:86` (@chatgpt-codex-connector) — [thread](https://github.com/nexu-io/looper/pull/306#discussion_r3231169443)
  - The daemon_boot paths filter now includes internal/bootstrap/** in .github/workflows/ci.yml, so bootstrap-only daemon startup changes still trigger the daemon boot E2E gate.
- ✅ Review comment on `.github/workflows/ci.yml:79` (@chatgpt-codex-connector) — [thread](https://github.com/nexu-io/looper/pull/306#discussion_r3231169447)
  - The all_e2e filter in .github/workflows/ci.yml now watches internal/e2e/*.go plus the harness directory, so shared top-level E2E helpers and entrypoints trigger the targeted E2E suites instead of bypassing them.


<sub>🔁 Powered by <a href="https://github.com/nexu-io/looper">Looper</a> · runner=fixer · agent=opencode · An autonomous AI dev team for your GitHub repos.</sub>

File: .github/workflows/release.yml [hunk 3]
```diff
@@ -34,7 +34,50 @@ env:
...
+
+      - name: Mint sandbox GitHub App token
+        id: app-token
+        uses: actions/create-github-app-token@v3
+        with:
+          app-id: ${{ vars.LOOPER_E2E_GITHUB_APP_ID }}
+          private-key: ${{ secrets.LOOPER_E2E_GITHUB_APP_PRIVATE_KEY }}
+          owner: nexu-io
+          repositories: |
+            looper-sandbox
```
[the remark this answers is not in the document]
Comment (mrcfps) [reply .github/workflows/release.yml]: @chatgpt-codex-connector This has been fixed in e5b67f2.

The release sandbox preflight now parses owner/repo from LOOPER_E2E_SANDBOX_REPO before minting the GitHub App token, so the token is scoped to the configured sandbox target in .github/workflows/release.yml.




<sub>🔁 Powered by <a href="https://github.com/nexu-io/looper">Looper</a> · runner=fixer · agent=opencode · An autonomous AI dev team for your GitHub repos.</sub>

File: .github/workflows/sandbox-cleanup.yml [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+    steps:
+      - name: Mint sandbox GitHub App token
+        id: app-token
+        uses: actions/create-github-app-token@v3
+        with:
+          app-id: ${{ vars.LOOPER_E2E_GITHUB_APP_ID }}
+          private-key: ${{ secrets.LOOPER_E2E_GITHUB_APP_PRIVATE_KEY }}
+          owner: nexu-io
+          repositories: |
+            looper-sandbox
```
[the remark this answers is not in the document]
Comment (mrcfps) [reply .github/workflows/sandbox-cleanup.yml]: @chatgpt-codex-connector This has been fixed in e5b67f2.

The sandbox cleanup workflow now resolves owner/repo from SANDBOX_REPO before creating the App token, so cleanup gh issue/pr commands stay authorized for the configured sandbox repo in .github/workflows/sandbox-cleanup.yml.




<sub>🔁 Powered by <a href="https://github.com/nexu-io/looper">Looper</a> · runner=fixer · agent=opencode · An autonomous AI dev team for your GitHub repos.</sub>

File: .github/workflows/ci.yml [hunk 5]
```diff
@@ -49,3 +49,245 @@ jobs:
...
+          filters: |
+            all_e2e:
+              - 'go.mod'
+              - 'go.sum'
+              - 'internal/e2e/harness/**'
+            daemon_boot:
+              - 'cmd/**'
+              - 'internal/config/**'
+              - 'internal/runtime/**'
+              - 'internal/e2e/smoke_daemon_test.go'
```
[the remark this answers is not in the document]
Comment (mrcfps) [reply .github/workflows/ci.yml:87]: @chatgpt-codex-connector This has been fixed in e5b67f2.

The daemon_boot paths filter now includes internal/bootstrap/** in .github/workflows/ci.yml, so bootstrap-only daemon startup changes still trigger the daemon boot E2E gate.




<sub>🔁 Powered by <a href="https://github.com/nexu-io/looper">Looper</a> · runner=fixer · agent=opencode · An autonomous AI dev team for your GitHub repos.</sub>

File: .github/workflows/ci.yml [hunk 6]
```diff
@@ -49,3 +49,245 @@ jobs:
...
+        if: github.event_name == 'pull_request'
+        id: filter
+        continue-on-error: true
+        uses: dorny/paths-filter@v3
+        with:
+          filters: |
+            all_e2e:
+              - 'go.mod'
+              - 'go.sum'
+              - 'internal/e2e/harness/**'
```
[the remark this answers is not in the document]
Comment (mrcfps) [reply .github/workflows/ci.yml:80]: @chatgpt-codex-connector This has been fixed in e5b67f2.

The all_e2e filter in .github/workflows/ci.yml now watches internal/e2e/*.go plus the harness directory, so shared top-level E2E helpers and entrypoints trigger the targeted E2E suites instead of bypassing them.




<sub>🔁 Powered by <a href="https://github.com/nexu-io/looper">Looper</a> · runner=fixer · agent=opencode · An autonomous AI dev team for your GitHub repos.</sub>

File: .github/workflows/ci.yml [hunk 7]
```diff
@@ -49,3 +49,247 @@ jobs:
...
+              - 'internal/reviewer/**'
+              - 'internal/api/**'
+              - 'internal/worktree*/**'
+              - 'internal/e2e/resolve_comments_scenarios_test.go'
+            worktree_invariants:
+              - 'internal/worker/**'
+              - 'internal/fixer/**'
+              - 'internal/reviewer/**'
+              - 'internal/api/**'
+              - 'internal/worktree*/**'
```
[the remark this answers is not in the document]
Comment (mrcfps) [reply .github/workflows/ci.yml:105]: @chatgpt-codex-connector This has been fixed in ac1b089.

Updated .github/workflows/ci.yml so the worktree_invariants paths filter also watches internal/infra/git/**, ensuring worktree-invariants-e2e still runs when git/worktree lifecycle code changes under internal/infra/git.




<sub>🔁 Powered by <a href="https://github.com/nexu-io/looper">Looper</a> · runner=fixer · agent=opencode · An autonomous AI dev team for your GitHub repos.</sub>

File: internal/bootstrap/bootstrap.go [hunk 8]
```diff
@@ -145,6 +150,39 @@ func signalNotifierOrDefault(notifier SignalNotifier) SignalNotifier {
...
+		if detection[check.statusKey] != config.ToolDetectionStatusConfigured || check.path == nil {
+			continue
+		}
+		value := strings.TrimSpace(*check.path)
+		if value == "" || (!filepath.IsAbs(value) && !strings.ContainsRune(value, os.PathSeparator)) {
+			continue
+		}
+		info, err := os.Stat(value)
+		if err == nil && !info.IsDir() {
+			if _, err := exec.LookPath(value); err == nil {
```
[the remark this answers is not in the document]
Comment (mrcfps) [reply internal/bootstrap/bootstrap.go:173]: @chatgpt-codex-connector This has been fixed in ca467f1.

Updated internal/bootstrap/bootstrap.go to only preflight absolute configured tool paths, so repo-relative wrappers are left to resolve from each command's runtime working directory. Added TestBootstrapAllowsConfiguredRelativeToolPath in internal/bootstrap/bootstrap_test.go to cover the regression.




<sub>🔁 Powered by <a href="https://github.com/nexu-io/looper">Looper</a> · runner=fixer · agent=opencode · An autonomous AI dev team for your GitHub repos.</sub>

**Comment by mrcfps** [comment]
**Looper fixer round complete** — ca467f1

- ✅ Review comment on `internal/bootstrap/bootstrap.go:173` (@chatgpt-codex-connector) — [thread](https://github.com/nexu-io/looper/pull/306#discussion_r3236224322)
  - Updated internal/bootstrap/bootstrap.go to only preflight absolute configured tool paths, so repo-relative wrappers are left to resolve from each command's runtime working directory. Added TestBootstrapAllowsConfiguredRelativeToolPath in internal/bootstrap/bootstrap_test.go to cover the regression.


<sub>🔁 Powered by <a href="https://github.com/nexu-io/looper">Looper</a> · runner=fixer · agent=opencode · An autonomous AI dev team for your GitHub repos.</sub>
